src/rust/session/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+//! This crate provides buffered TCP sessions with or without TLS which can be
+//! used with [`::mio`]. TLS/SSL is provided by BoringSSL with the [`::boring`]
+//! crate.
+
+#[macro_use]
+extern crate rustcommon_fastmetrics;
+
+mod buffer;
+mod stream;
+mod tcp_stream;
+
+use std::borrow::Borrow;
+use std::io::{ErrorKind, Read, Write};
+use std::net::SocketAddr;
+
+use boring::ssl::{MidHandshakeSslStream, SslStream};
+use bytes::Buf;
+use common::traits::ExtendFromSlice;
+use metrics::Stat;
+use mio::event::Source;
+use mio::{Interest, Poll, Token};
+
+use buffer::Buffer;
+use stream::Stream;
+
+pub use tcp_stream::TcpStream;
+
+pub const MIN_BUFFER_SIZE: usize = 1024; // 1 KiB
+
+// TODO(bmartin): implement connect/reconnect so we can use this in clients too.
+/// The core `Session` type which represents a TCP stream (with or without TLS),
+/// the session buffer, the mio [`::mio::Token`],
+pub struct Session {
+    token: Token,
+    // TODO(bmartin): remove this lint exception after implementing connect
+    #[allow(dead_code)]
+    addr: SocketAddr,
+    stream: Stream,
+    read_buffer: Buffer,
+    write_buffer: Buffer,
+    tmp_buffer: [u8; MIN_BUFFER_SIZE],
```

Comment:
Should this be hidden from the public trait interface?

Reply:
Yes. I need to see why I leaked that when splitting the code into crates. We probably want to pass that value as part of session initialization.

Reply:
This is now passed during session creation and defined in rust/core/server